Data scraping with selenium from [ryans](https://www.ryanscomputers.com/category/laptop-all-laptop)

Installing and loading required libraries.

In [ ]:
%pip install ChromeDriverManager


In [30]:
from datetime import date
today = str(date.today())
print(today)

2023-12-27


In [1]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
import time


Here first, we are checking for one url

In [2]:
option = webdriver.ChromeOptions()
driver = webdriver.Chrome(service = Service(ChromeDriverManager().install()),options=option)

In [3]:
driver.get("https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0")

We take second page for finding the pattern for pagination

https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page=2

We are trying to make similar  patters for all page

In [3]:
#pagination
for page in range(1,5):
    url = f'https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page={page}'
    print(url)

https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page=1
https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page=2
https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page=3
https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page=4


Here, we are tyring to get specific url of each laptop.

In [4]:
links = driver.find_elements('xpath', '//p[@class="card-text p-0 m-0 grid-view-text"]/a')


In [5]:
for link in links:
    print(link.get_attribute('href'))

https://www.ryanscomputers.com/lenovo-ideapad-d330-10igl-intel-cdc-n4020-mineral-grey-laptop
https://www.ryanscomputers.com/hp-15s-fq3617tu-intel-cdc-n4500-15.6-inch-fhd-display-silver-laptop
https://www.ryanscomputers.com/hp-15s-eq1578au-amd-athlon-silver-3050u-15.6-inch-fhd-display-silver-laptop
https://www.ryanscomputers.com/lenovo-ideapad-1-15ada7-amd-ryzen-3-3250u-15.6-inch-fhd-display-cloud-grey-laptop
https://www.ryanscomputers.com/lenovo-ideapad-slim-3i-15itl-11th-gen-intel-core-i3-1115g4-15.6-inch-fhd-display-arctic-grey-laptop
https://www.ryanscomputers.com/acer-aspire-lite-al15-51-intel-core-i3-1115g4-8gb-15.6-inch-fhd-display-titanium-gray-laptop
https://www.ryanscomputers.com/dell-inspiron-15-3515-amd-ryzen-3-3250u-15.6-inch-fhd-display-silver-laptop
https://www.ryanscomputers.com/hp-15s-eq2618au-amd-ryzen-3-5300u-15.6-inch-hd-display-spruce-blue-laptop
https://www.ryanscomputers.com/hp-15s-du3611tu-intel-core-i3-1125g4-15.6-inch-fhd-display-jet-black-laptop
https://www.ry

This are the all laptop pages links of first page.

In [6]:
len(links)

100

We are creating list of all laptops of all(4) page

In [7]:
#pagination
urls = []
for page in range(1,5):
    url = f'https://www.ryanscomputers.com/category/laptop-all-laptop?limit=100&osp=1&st=0&page={page}'
    driver.get(url)
    time.sleep(3)
    links = driver.find_elements('xpath', '//p[@class="card-text p-0 m-0 grid-view-text"]/a')
    for link in links:
        urls.append(link.get_attribute('href'))

In [8]:
len(urls)

373

WE can see that till date we have 373 laptops and its urls.  It changes actually as they updates on their website.

Now we are trying to find the data from one laptop.

In [9]:
driver.get(urls[0])

We tried to extract some data such as title, regular price, speial price, some features etc.

In [10]:
title = driver.find_element('xpath','//h1[@itemprop="name"]').text
title

'Lenovo IdeaPad D330 10IGL Intel CDC N4020 4GB RAM 128GB eMMC 10.1 Inch HD IPS Touch Display Mineral Grey Laptop'

In [11]:
regular_price = int(driver.find_element('xpath','//span[@class="rp-block mb-2"]/span[contains(text(),"Regular Price")]').text.replace('Regular Price Tk ','').replace(',',''))
regular_price

38630

In [12]:
special_price = int(driver.find_element('xpath','//span[@class="sp-block"]').text.replace('Special Price Tk ','').replace(',',''))
special_price

35800

In [13]:
monthly_emi = int(driver.find_element('xpath','//a[contains(text(),"Monthly EMI")]').text.replace('Monthly EMI Tk','').replace(',',''))
monthly_emi

3219

In [14]:
all_featurs = driver.find_elements('xpath','//span[@class="att-title context"]') 


In [15]:
for index, feature in enumerate(all_featurs,1):
    print(index,feature.text)

1 Brand
2 Model
3 Laptop Series
4 Part No
5 Processor Brand
6 Processor Type.
7 Generation
8 Processor Model
9 Processor Base Frequency
10 Processor Max Turbo Frequency
11 Processor Core
12 Processor Thread
13 CPU Cache
14 Chipset.
15 RAM
16 Installed RAM Details
17 RAM Type
18 RAM Bus (MHz)
19 Total RAM Slot
20 Empty/Expansion RAM Slot
21 Max. RAM Support
22 Optane Memory
23 Optane Memory Slot
24 Storage
25 Installed HDD Type
26 HDD RPM
27 HDD Expansion Slot
28 Installed SSD Type
29 M.2/SSD Expansion Slot
30 Storage Upgrade
31 Graphics Chipset
32 Graphics Memory Accessibility
33 Graphics Memory
34 Graphics Memory Type
35 Display Size (Inch)
36 Display Type
37 Display Resolution
38 Display Surface
39 Touch Screen
40 Display Refresh Rate
41 Display Bezel
42 Display Features
43 Secondary Screen Size
44 Secondary Screen Technology
45 Secondary Screen Resolution
46 Optical Drive
47 Multimedia Card Slot
48 Supported Multimedia Card
49 USB 2 Port
50 USB 3 Port
51 USB C / Thunderbolt Port
52 

In [16]:
for index, value in enumerate(all_featurs,1):
    if value.text =='Brand':
        brand = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]')
        print(brand.text)
    elif value.text =='Model':
        model = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]')
        print(model.text)
    elif value.text =='Storage':
        try:
            storage = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]')
            print(storage.text)  
        except:
            storage = "NA"      

Lenovo
Lenovo IdeaPad D330 10IGL
128GB eMMC


Among lots of features we extract brand, model and storage for practice purpose.

Now we are doing the same process for all laptops.

In [ ]:
data=[]
for url in urls:
    driver.get(url)
    time.sleep(2)
    try:
        title = driver.find_element('xpath','//h1[@itemprop="name"]').text
    except:
        title = 'NA'
    try:
        reg_price = int(driver.find_element('xpath','//span[@class="rp-block mb-2"]/span[contains(text(),"Regular Price")]').text.replace('Regular Price Tk ','').replace(',',''))
    except:
        reg_price = 'NA'
    try:
        spe_price = int(driver.find_element('xpath','//span[@class="sp-block"]').text.replace('Special Price Tk ','').replace(',',''))
    except:
        spe_price = int(driver.find_element('xpath','//span[@class="sp-block"]/span[contains(text(),"Tk")]').text.replace('Tk ','').replace(',',''))   
    
    emi = int(driver.find_element('xpath','//a[contains(text(),"Monthly EMI")]').text.replace('Monthly EMI Tk','').replace(',',''))
    all_featurs = driver.find_elements('xpath','//span[@class="att-title context"]') 
    for index, value in enumerate(all_featurs,1):
        if value.text =='Brand':
            brand = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]').text
        elif value.text =='Model':
            model = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]').text
        elif value.text =='Storage':
            try:
                storage = driver.find_element('xpath',f'(//span[@class="att-value context"])[{index}]').text
            except:
                storage = 'NA'
    d = {
        'title' : title,
        'regukar_price' : reg_price,
        'special_price' : spe_price,
        'monthly_emi': emi,
        'brand' : brand,
        'model': model,
        'storage' : storage
    }  
    data.append(d)           

After 257 laptops there we faced  problems with `spe-price`, xpath changed or class name changed.  But for now its good we got 257 laptops data.

In [21]:
import pandas as pd

In [26]:
d=pd.DataFrame(data)

In [27]:
d.head(80)

title  regukar_price  \
0   Lenovo IdeaPad D330 10IGL Intel CDC N4020 4GB ...          38630   
1   HP 15s-fq3617TU Intel CDC N4500 4GB RAM 256GB ...          46110   
2   HP 15s-eq1578AU AMD Athlon Silver 3050U 8GB 25...          51540   
3   Lenovo IdeaPad 1 15ADA7 AMD Ryzen 3 3250U 8GB ...          53930   
4   Lenovo IdeaPad Slim 3i 15ITL 11th Gen Intel Co...          55340   
..                                                ...            ...   
75  Lenovo IdeaPad 1 15AMN7 AMD Ryzen 3 7320U 8GB ...          57510   
76  Lenovo IdeaPad 1 15AMN7 AMD Ryzen 3 7320U 8GB ...          57510   
77  Lenovo IdeaPad Slim 3i 15ITL Intel Core i3 111...          58920   
78  MSI Modern 14 C11M Intel Core i3 1115G4 8GB RA...          59130   
79  Asus X415EA Intel Core i3 1115G4 4GB RAM 256GB...          59680   

    special_price  monthly_emi   brand                         model  \
0           35800         3219  Lenovo     Lenovo IdeaPad D330 10IGL   
1           42500         3843      HP               HP 15s-fq3617TU   
2           47500         4295      HP               HP 15s-eq1578AU   
3           49900         4494  Lenovo       Lenovo IdeaPad 1 15ADA7   
4           51400         4612  Lenovo  Lenovo IdeaPad Slim 3i 15ITL   
..            ...          ...     ...                           ...   
75          53800         4793  Lenovo       Lenovo IdeaPad 1 15AMN7   
76          53800         4793  Lenovo       Lenovo IdeaPad 1 15AMN7   
77          54500         4910  Lenovo  Lenovo IdeaPad Slim 3i 15ITL   
78          55000         4928     MSI            MSI Modern 14 C11M   
79          55000         4973    Asus                   Asus X415EA   

       storage  
0   128GB eMMC  
1    256GB SSD  
2    256GB SSD  
3    512GB SSD  
4    256GB SSD  
..         ...  
75   512GB SSD  
76   512GB SSD  
77   512GB SSD  
78   512GB SSD  
79   256GB SSD  

[80 rows x 7 columns]

In [28]:
len(d)

257

In [29]:
d.to_csv('laptops.csv',index = False)